### PR TITLE
[qt5-base, qt5-imageformat] fix issues on osx 

### LINKED ISSUE #6733: [Qt5] replace fixcmake.py with a cmake script
because it is quite similar to vcpkg_fixup_cmake_targets + some additional steps to correct the cmake files.

### LINKED ISSUE #9299: [qt5-xmlpatterns/qt5-tools] needs deployment of platform plugins into its tools dir
otherwise they cannot be used

So simply copy `plugins/platform` to `tools/qt5-xmlpatterns/bin/plugins/platform` solves the problem
(is required by paraview)

### LINKED ISSUE #10173: [qt5:x64-osx] build failure
**Host Environment**
- OS: [MacOS Catalina 10.15.3]
- Compiler: Apple clang version 11.0.0 (clang-1100.0.33.17)

**To Reproduce**
`./vcpkg install qt5:x64-osx`

**Failure logs** 
-Cut and past the appropriate build messages from the console output.
```
Building package qt5-imageformats[core]:x64-osx...
-- Using cached /Users/dean/Projects/CLionProjects/vcpkg/downloads/qtimageformats-everywhere-src-5.12.5.tar.xz
-- Using source at /Users/dean/Projects/CLionProjects/vcpkg/buildtrees/qt5-imageformats/src/5.12.5-7ddf5dd29e
-- Configuring x64-osx-rel
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: /Users/dean/Projects/CLionProjects/vcpkg/installed/x64-osx/tools/qt5/bin/qmake CONFIG-=debug CONFIG+=release CONFIG-=shared CONFIG*=static /Users/dean/Projects/CLionProjects/vcpkg/buildtrees/qt5-imageformats/src/5.12.5-7ddf5dd29e -qtconf /Users/dean/Projects/CLionProjects/vcpkg/buildtrees/qt5-imageformats/x64-osx-rel/qt.conf -- -system-tiff -system-webp -jasper -no-mng -verbose "TIFF_LIBS=/Users/dean/Projects/CLionProjects/vcpkg/installed/x64-osx/lib/libtiff.a /Users/dean/Projects/CLionProjects/vcpkg/installed/x64-osx/lib/liblzma.a" "WEBP_LIBS=/Users/dean/Projects/CLionProjects/vcpkg/installed/x64-osx/lib/libwebp.a /Users/dean/Projects/CLionProjects/vcpkg/installed/x64-osx/lib/libwebpdemux.a" "JASPER_LIBS=/Users/dean/Projects/CLionProjects/vcpkg/installed/x64-osx/lib/libjasper.a FREEGLUT_RELEASE-NOTFOUND"
    Working Directory: /Users/dean/Projects/CLionProjects/vcpkg/buildtrees/qt5-imageformats/x64-osx-rel
    Error code: 3
    See logs for more information:
      /Users/dean/Projects/CLionProjects/vcpkg/buildtrees/qt5-imageformats/config-x64-osx-rel-out.log
      /Users/dean/Projects/CLionProjects/vcpkg/buildtrees/qt5-imageformats/config-x64-osx-rel-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_qmake.cmake:61 (vcpkg_execute_required_process)
  installed/x64-osx/share/qt5/qt_build_submodule.cmake:10 (vcpkg_configure_qmake)
  installed/x64-osx/share/qt5/qt_submodule_installation.cmake:8 (qt_build_submodule)
  ports/qt5-imageformats/portfile.cmake:37 (qt_submodule_installation)
  scripts/ports.cmake:90 (include)


Error: Building package qt5-imageformats:x64-osx failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: qt5-imageformats:x64-osx
  Vcpkg version: 2020.02.04-unknownhash

Additionally, attach any relevant sections from the log files above.
```
-Please attach any additional failure logs mentioned in the console output.

**Additional context**
Add any other context about the problem here, such as what you have already tried to resolve the issue.

### PR DESCRIPTION
**Pull request by Neumann-A**
continuation of #9686 (broken commit history)
~~requires~~ includes #9680 (to make CI work)
 
qt5-base: added iconv dependency to fontconfig
qt5-imageformat: removed freeglut dependency on osx

closes #6733
closes #9299
duplicate
closes #10173

### CODE REVIEW

**Comment by JackBoosY** [comment]
Seems some ports can be successfully built, but some ports are broken.

**Comment by Neumann-A** [comment]
I'll fix the remaining regressions when i got the time for it. I already know how to fix those due to my local second paraview branch with qt 5.14 but have to apply those changes to Qt 5.12

**Comment by Neumann-A** [comment]
Ah nice .... wt has a silent dependency on Qt5 ....

**Review by JackBoosY — APPROVED** [review]

**Review by PhoebeHui — APPROVED** [review]

File: ports/jasper/CONTROL [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Source: jasper
-Version: 2.0.16-2
+Version: 2.0.16-3
 Homepage: https://github.com/mdadams/jasper
 Description: Open source implementation of the JPEG-2000 Part-1 standard
-Build-Depends: libjpeg-turbo, opengl, freeglut
+Build-Depends: libjpeg-turbo, opengl, freeglut (!osx)
```
Comment (vicroms) [inline ports/jasper/CONTROL]: This has been updated on master #9680

Comment (Neumann-A) [reply ports/jasper/CONTROL]: removed by the merge with master

File: scripts/ci.baseline.txt [hunk 2]
```diff
@@ -1,1836 +1,1833 @@
...
-yato:arm64-windows=fail
-yato:arm-uwp=fail
-yato:x64-uwp=fail
-z3:arm64-windows=fail
-z3:arm-uwp=fail
-z3:x64-uwp=fail
-zeromq:arm64-windows=fail
-zeromq:arm-uwp=fail
-zeromq:x64-uwp=fail
+###########################################################################
```
Comment (Neumann-A) [inline scripts/ci.baseline.txt]: Do I need to understand why the web editor chose to completely change the file on the merge.....?

**Comment by JackBoosY** [comment]
Please update ci.baseline.txt, thanks.

**Comment by david-antiteum** [comment]
10 days with no Qt5. Vcpkg without versioning is unusable (or I'm missing something)

**Comment by Neumann-A** [comment]
> or I'm missing something

vcpkg overlay ports + git history?

**Comment by david-antiteum** [comment]
> vcpkg overlay ports + git history?

Thanks @Neumann-A and sorry to bring this issue here, but, as we are now talking let me dig a bit more. In Conan you can setup a requirements file as:

[requires]
Poco/1.9.0@pocoproject/stable
zlib/1.2.11@conan/stable

How can this be done in vcpkg? In particular, when I´m setting up a new computer in where you will start with a fresh installation of vcpkg (with the latest ports).

Thanks a lot!

File: ports/qt5-base/portfile.cmake [hunk 3]
```diff
@@ -299,12 +301,28 @@ else()
...
     endif()
 
-    #TODO: Replace python script with cmake script
-    vcpkg_execute_required_process(
-        COMMAND ${PYTHON3} ${CMAKE_CURRENT_LIST_DIR}/fixcmake.py
-        WORKING_DIRECTORY ${CURRENT_PACKAGES_DIR}/share/cmake
-        LOGNAME fix-cmake
-    )
+    ## Fix location of qtmain(d).lib. Has been moved into manual-link. Add debug version
+    if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_BUILD_TYPE)
```
Comment (ras0219-msft) [inline ports/qt5-base/portfile.cmake]: Does this not need adjustment for non-windows platforms?

File: scripts/ci.baseline.txt [hunk 4]
```diff
@@ -1404,23 +1403,22 @@ python3:x64-uwp=fail
...
-qscintilla:x64-osx=fail
-qt5-activeqt:x64-linux=fail
-qt5-activeqt:x64-osx=fail
-qt5-macextras:x64-linux=fail
-qt5-macextras:x64-windows=fail
-qt5-macextras:x64-windows-static=fail
-qt5-macextras:x86-windows=fail
-qt5-winextras:x64-linux=fail
-qt5-winextras:x64-osx=fail
+#unsupported platforms/systems
```
Comment (ras0219-msft) [inline scripts/ci.baseline.txt]: We could use the new `Supports:` field for this, but don't worry about putting it into this PR (we'll do a cleanup pass in the future once the new functionality is fully stabilized).

**Comment by ras0219-msft** [comment]
@david-antiteum Vcpkg manages versions via git commits; so there are several ways to easily pin every version like you're describing:

1. git submodule (this ties commits of the project to commits of vcpkg, which works really well)
2. A readme.txt file noting the "last known good" vcpkg commit id
3. Fork us on github and use master/branches/tags to manage what package sets you want to use (this is great for the new computer scenario)

To mix versions across different vcpkg commits, which we highly recommend against because they are not tested, you can use a combination of the techniques above plus overlay ports.

We are looking at a "manifest file" similar to what you described, however there are quite a few edge cases that need to be handled to have a good experience.

**Comment by david-antiteum** [comment]
@ras0219-msft Thanks for the explanation. I did some experiments using the 3rd option a few months ago (see http://www.antiteum.com/2019/05/vcpkg-versioning.html) and I end with a DB of packages and versions after scanning the complete git history. Anyway, I will wait for the manifest option :)

File: ports/qt5-base/portfile.cmake (hunk 3 above)
Comment (Neumann-A) [reply ports/qt5-base/portfile.cmake]: No Qt5::WinMain only exists on Windows. According to `Qt5CoreConfigExtras.cmake.in` other platforms do not have `<Os>Main` a target

**Comment by Neumann-A** [comment]
@JackBoosY: Any idea why osg-qt fails? Seems to be a more general problem since I saw a similar failure in other build logs.

**Comment by JackBoosY** [comment]
`osg-qt` has some file duplication issues with other ports, so I disable it in #10290.
I'm not sure if this is related to this PR.

**Comment by JackBoosY** [comment]
Due to the conflict between qt5-base and ms-angle, libEGL (depending on ms-angle) generation failed, which eventually caused this issue.

**Comment by Neumann-A** [comment]
> Due to the conflict between qt5-base and ms-angle, libEGL (depending on ms-angle) generation failed, which eventually caused this issue.

qt5-base no longer builds angle so there is no conflict there.
The conflict is between angle and ms-angle but since ms-angle is set to skip it should load all libraries build by angle unless angle does not build libEGL (which it should).

**Comment by Neumann-A** [comment]
@JackBoosY: Should we jump directly to #10009 if possible. I just split the PRs up because I thought the fixes would land faster in master. (And i didn't now when i will finish them.)

**Comment by JackBoosY** [comment]
@Neumann-A Okay, that's ok.

**Comment by Neumann-A** [comment]
...... ah really i should not use the github merge tool anymore. It seems to completely replace the file all the time....

**Comment by JackBoosY** [comment]
Retesting...

**Review by JackBoosY — APPROVED** [review]

**Comment by heydojo** [comment]
Hi All,
This merge fails for me in the following way:

`Starting package 16/16: qt5-base:x64-windows
Building package qt5-base[core,latest]:x64-windows...
-- Figuring out qt target mkspec. Target arch x64
-- Target mkspec set to: win32-msvc
-- Host mkspec not set. Qt's own buildsystem will try to figure out the host system
-- Downloading http://download.qt.io/official_releases/qt/5.13/5.13.1/submodules/qtbase-everywhere-src-5.13.1.tar.xz...
-- Extracting source H:/vcpkg/downloads/qtbase-everywhere-src-5.13.1.tar.xz
-- Applying patch patches/winmain_pro.patch
-- Applying patch patches/windows_prf.patch
-- Applying patch patches/qt_app.patch
-- Applying patch patches/gui_configure.patch
-- Applying patch patches/icu.patch
-- Applying patch patches/Qt5BasicConfig.patch
CMake Error at scripts/cmake/vcpkg_apply_patches.cmake:51 (message):
  Applying patch failed.  Checking patch
  mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in...

  Hunk #1 succeeded at 124 (offset 71 lines).

  error: while searching for:

      )



  !!IF !isEmpty(CMAKE_WINDOWS_BUILD)

  !!IF isEmpty(CMAKE_LIB_DIR_IS_ABSOLUTE)

      set(imported_implib \"${_qt5$${CMAKE_MODULE_NAME}_install_prefix}/$${CMAKE_LIB_DIR}${IMPLIB_LOCATION}\")

  !!ELSE

      set(imported_implib \"IMPORTED_IMPLIB_${Configuration}\" \"$${CMAKE_LIB_DIR}${IMPLIB_LOCATION}\")

  !!ENDIF



  error: patch failed: mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in:69

  error: mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in: patch does not
  apply

Call Stack (most recent call first):
  scripts/cmake/vcpkg_extract_source_archive_ex.cmake:134 (vcpkg_apply_patches)
  ports/qt5-base/cmake/qt_download_submodule.cmake:30 (vcpkg_extract_source_archive_ex)
  ports/qt5-base/portfile.cmake:35 (qt_download_submodule)
  scripts/ports.cmake:90 (include)


Error: Building package qt5-base:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: qt5-base:x64-windows
  Vcpkg version: 2020.02.04-nohash-external

Additionally, attach any relevant sections from the log files above.`

--------------------------------------------------------------------

To test, I did the following:

- Cloned https://github.com/Neumann-A/vcpkg
- Switched to feature branch **fix_osx_qt_2** (git checkout fix_osx_qt_2)
- Moved into the vcpkg directory and ran `CALL .\bootstrap-vcpkg.bat`
- Issued `vcpkg install qt5-base[latest]:x64-windows`

This resulted in the build failure above.
Hope this helps.

**Comment by Neumann-A** [comment]
@heydojo: 
This one is only intended to fix 5.12.5 (osx). 
I know that the changes here are breaking the [latest] feature but since I am already updating latest to 5.14.2 (#10644) I am not going to spend time to fix this PR for 5.13.1 (If you really need 5.13.1 the patches need to be updated since they contain the magic to make single and multi configuration builds work correctly). Currently you could try out #10009 (5.14.1) until I have the patches fixed for 5.14.2 in #10644

**Comment by JackBoosY** [comment]
Hi @Neumann-A, please resolve baseline conflicts.

Thanks.

**Comment by heydojo** [comment]
@Neumann-A Thanks. Please disregard my message regarding the patches here. I did try your 5.14.1 branch and will test the 5.14.2 branch you have with the qtmain.lib fix https://github.com/microsoft/vcpkg/pull/10644

**Comment by Neumann-A** [comment]
@JackBoosY: Any plans to merge this? The only reason I keep this open to be merge is to have a working qt 5.12.5 version for osx in the commit history of vcpkg. Otherwise I could simply close this in favor of the other Qt PR.

**Comment by JackBoosY** [comment]
@ras0219-msft @dan-shaw Could you please merge this PR first？

**Review by dan-shaw — APPROVED** [review]

**Comment by dan-shaw** [comment]
@Neumann-A @JackBoosY LGTM

**Pull request merged**